plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -3710,57 +3710,79 @@ public void addPartition(PartitionWithStatistics partition)
             partitions.add(partition);
         }
 
-        public void execute(AcidTransaction transaction)
+        public void execute(AcidTransaction transaction, Executor executor)
         {
+            List<CompletableFuture<List<List<String>>>> futures = new ArrayList<>();
             List<List<PartitionWithStatistics>> batchedPartitions = Lists.partition(partitions, batchSize);
             for (List<PartitionWithStatistics> batch : batchedPartitions) {
+                futures.add(CompletableFuture.supplyAsync(() -> this.executeForBatch(batch), executor));
+            }
+
+            try {
+                futures.forEach(future -> createdPartitionValues.addAll(getFutureValue(future, TrinoException.class)));
+            }
+            catch (Throwable t) {
                 try {
-                    metastore.addPartitions(schemaName, tableName, batch);
-                    for (PartitionWithStatistics partition : batch) {
-                        createdPartitionValues.add(partition.getPartition().getValues());
-                    }
+                    futures.forEach(future -> future.cancel(true));
                 }
-                catch (Throwable t) {
-                    // Add partition to the created list conservatively.
```

Comment:
does it actually go though this `catch` in benchmarks?

Instead of using ` metastore.getPartition` we could use `getPartitionsByNames` which is batched.

Reply:
> does it actually go though this catch in benchmarks?

no